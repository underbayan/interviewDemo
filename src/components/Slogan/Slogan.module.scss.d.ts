export const descriptionStyle: string;
export const sloganStyle: string;
export const titleStyle: string;
