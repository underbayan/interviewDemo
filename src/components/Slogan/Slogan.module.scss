.sloganStyle {
  text-transform: uppercase;
  color: #888;
  display: flex;
  flex-direction: column;

  .titleStyle {
    font-size: 3rem;
    margin: 0;
    font-weight: 600;
    font-family: 'Times New Roman', Times, serif;
  }

  .descriptionStyle {
    margin: 20px 0 0 0;
    font-size: 1rem;
    align-self: flex-end;
  }
}