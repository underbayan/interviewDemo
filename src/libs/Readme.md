"Libs" refer to the libraries or modules that contain code that is meant to be reusable across different parts of an application. In the context of the specified constraint, the code in "libs" should not contain any mutations logic. Rather, it should be focused on providing functionalities that are shared across different parts of the application without altering any data. For instance, "libs" could contain helper functions, utility classes, or any other non-mutating code that can be used in different parts of the application.