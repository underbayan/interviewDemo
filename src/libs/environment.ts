import stringsEn from "../mock.strings.en.json"
import stringsZh from "../mock.strings.zh.json"

export const environment = {
  app: "demo",
  page: "/",
  lang: "en",
  version: "0.0.1",
  flights: "", // AB test
  config: {},  // Config for page
  strings: stringsEn,
  langMap: {
    en: stringsEn,
    zh: stringsZh
  }
  // TODO: i18n.detectLanguage() to get zh/en/fr lang config json
}