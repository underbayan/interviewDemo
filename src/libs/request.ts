export interface FetchDataType {
  [key: string]: any
}

export interface AbortableFetchResponse extends Response {
  data?: FetchDataType | string | ArrayBuffer | Blob
  message: string
}

export interface AbortableFetchOptions extends RequestInit {
  abort?: AbortSignal
}

export interface AbortableFetchPromise extends Promise<AbortableFetchResponse> {
  [AbortableKey](): void
}

export const AbortableKey = Symbol.for('abortableKey')
export const abortableFetch = (
  url: string,
  options?: AbortableFetchOptions
): AbortableFetchPromise => {
  const { abort, ...fetchOptions } = options ?? {}
  const ab = new AbortController()
  const rt = (fetch(url, { ...fetchOptions, signal: abort ? abort : ab.signal }) as Promise<Response>)
    .then(async (response) => {
      const contentType = response.headers.get('content-type')
      let data: FetchDataType | undefined | string

      if (contentType?.includes('application/json')) {
        data = await response.json()
      } else if (contentType?.includes('application/text')) {
        data = await response.text()
      } else if (contentType?.includes('application/opaque')) {
        data = await response.blob()
      } else {
        data = await response.arrayBuffer()
      }

      const result: AbortableFetchResponse = {
        ...response,
        status: response.status,
        message: response.statusText,
        data,
      }
      return result
    })
    .catch((e) => console.error(e)) as AbortableFetchPromise

  Object.defineProperty(rt, AbortableKey, { value: () => ab.abort(), writable: false })

  return rt
}
