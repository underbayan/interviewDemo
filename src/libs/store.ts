import { createSlice, configureStore } from '@reduxjs/toolkit'
import { useSelector, useDispatch } from 'react-redux'
import { environment } from "./environment"
import { Dispatch, AnyAction } from 'redux'
export type RootState = ReturnType<typeof store.getState>

const envSlice = createSlice({
  name: 'env',
  initialState: environment,
  reducers: {

    updateEnv(state, action) {

      const newState = { ...state, ...action?.payload }
      if (action?.payload?.lang) newState.strings = newState.langMap[action.payload?.lang]
      return newState
    }
  }
})
export const store = configureStore({
  reducer: {
    env: envSlice.reducer
  },
})
interface EnvUpdatePayload {
  lang?: string
  [key: string]: any
}




export const useI18n = () => useSelector((state: RootState) => state.env.strings)
export const useEnv = () => useSelector((state: RootState) => state.env)
export const useUpdateEnv = (): ((env: EnvUpdatePayload) => void) => {
  const dispatch = useDispatch()

  const updateEnvAsync = (env: EnvUpdatePayload) => (dispatch: Dispatch) => {
    setTimeout(() => { dispatch(envSlice.actions.updateEnv(env)) }, 15)
  }

  return (env: EnvUpdatePayload) => dispatch(updateEnvAsync(env) as unknown as AnyAction)
}


