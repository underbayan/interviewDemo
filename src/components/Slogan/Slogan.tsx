import { useI18n } from "../../libs/store"
import { memo } from 'react'

import { sloganStyle, titleStyle, descriptionStyle } from "./Slogan.module.scss"
export const Slogan = memo(() => {
  const str = useI18n()
  return (<section className={sloganStyle}>
    <pre className={titleStyle}>{str.slogan}</pre>
    <p className={descriptionStyle}>{str.sloganDetail}</p>
  </section >)
})



