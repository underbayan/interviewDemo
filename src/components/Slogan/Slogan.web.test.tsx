import React from 'react';
import { render } from '@testing-library/react';
import { test, expect } from 'vitest';
import { Slogan } from './Slogan';
import { store } from "@demo/libs/store"
import { Provider } from 'react-redux'

test('Header renders correctly', () => {
  const { container } = render(<Provider store={store}>
    <Slogan></Slogan>
  </Provider>);
  expect(container).toMatchSnapshot();
});
