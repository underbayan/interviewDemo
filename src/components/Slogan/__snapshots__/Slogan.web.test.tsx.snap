// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header renders correctly 1`] = `
<div>
  <section>
    <pre>
      A better way 
To enjoy every day.
    </pre>
    <p>
      Be the first to know  when we launch
    </p>
  </section>
</div>
`;
